### PR TITLE
Support configuration of protocol binding for authentication requests

### PR DESCRIPTION
**Pull request by blucas**
Fixes gh-7513.

Note: This is not intended as the final solution, but as a starting point for discussion.  I believe a better solution may be to add `protocolBinding` to `Saml2AuthenticationRequest` and `RelyingPartyRegistration`.  But I'd like to hear feedback on that approach before submitting any code.

@fhanik FYI.

### CODE REVIEW

**Comment by fhanik** [comment]
Awesome @blucas 

Can you add a test case to your pull request, you can use this one so you get familiar with our style:
https://github.com/fhanik/spring-security/commit/f195ac7f67cb2d4a5a28ab5006121c0f92bea816

The naming convention for methods is
1. method name of method being tested
2. 'When'
3. condition for tests
4. 'Then'
5. expected results

**Comment by blucas** [comment]
@fhanik - I didn't add a test as I wasn't convinced this approach was the best solution to the problem.  Do you have any thoughts around exposing the `protocolBinding` to `Saml2AuthenticationRequest` and `RelyingPartyRegistration`?

**Comment by blucas** [comment]
@fhanik - I've added the test and fixed the failure.  Please consider my proposal around exposing the `protocolBinding` to `Saml2AuthenticationRequest` and `RelyingPartyRegistration`

**Comment by fhanik** [comment]
Hi @blucas ,

I've spent some time thinking about this. And the different options as you noted are

1. Just default to POST, no configuration option
2. Allow the `OpenSamlAuthenticationRequestFactory` to be configured for POST/REDIRECT
3. Expose it in the `Saml2AuthenticationRequest` object
4. Expose it in the `RelyingPartyRegistration` object
5. Make is available in Spring Boot's application.yml file

The further down the list we reach, the broader the impact it has on the overall codebase.

Then I started thinking about the use case. It is my belief that `POST` will probably cover 95%+ of all use cases. I can see why `REDIRECT` may be frowned upon because you're risking an assertion to be shared elsewhere through the browser, for example via the `Referrer` header.

I believe that option 1 and 2 are both sufficient. I believe going further down the list, why it may seem flexible, it's a lot of work for very little gain. Specially if we don't see this value change frequently. So I'm thinking a minimalist approach for the sake of simplicity and not creating an overarching configuration base.

What are your thoughts? Am I missing something ?

**Comment by blucas** [comment]
@fhanik I completely agree with everything you said.  I do believe `POST` as default is the right direction and covers majority of use cases.  Additionally it doesn't expose the SAML object to other services (such as access logs).

I feel like the `setter` approach provided in this PR is the bare minimum we can propose and will suffice for my purposes at the very least.

The reason I mention exposing it further (3 and 4 above) is it feels more like the "spring way" of doing things.  Plus is provides more specific configuration (one RPR could use `POST` while another could use `REDIRECT`).  But having said that, I don't think many people would make sure of that configurability/flexibility.

In summary.  I believe the approach of this PR is enough.  Any further configuration requests could be handled at a later date.

File: saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSamlAuthenticationRequestFactory.java [hunk 1]
```diff
@@ -67,4 +70,18 @@ public void setClock(Clock clock) {
 		Assert.notNull(clock, "clock cannot be null");
 		this.clock = clock;
 	}
+
+	/**
+	 * Sets the {@code protocolBinding} to use when generating authentication requests
+	 *
```
Comment (fhanik) [inline saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSamlAuthenticationRequestFactory.java]: Can you add to this documentation
1. What values are acceptable
2. @throws for when invalid parameter is set

I wonder if we should switch to `IllegalArgumentException` since this is not really SAML, but just a simple getter that takes a value.

File: saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/authentication/OpenSamlAuthenticationRequestFactoryTests.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	@Test
+	public void createAuthenticationRequestWhenSetUriThenReturnsCorrectBinding() {
+		factory.setProtocolBinding(SAMLConstants.SAML2_REDIRECT_BINDING_URI);
+		AuthnRequest authn = getAuthNRequest();
+		Assert.assertEquals(SAMLConstants.SAML2_REDIRECT_BINDING_URI, authn.getProtocolBinding());
+	}
+
+	@Test
+	public void createAuthenticationRequestWhenSetUnsupportredUriThenThrowsSaml2Exception() {
+		exception.expect(Saml2Exception.class);
```
Comment (fhanik) [inline saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/authentication/OpenSamlAuthenticationRequestFactoryTests.java]: Is `IllegalArgumentException` perhaps better

**Comment by fhanik** [comment]
Thank @blucas , I'll check out the CI failures tomorrow

**Comment by blucas** [comment]
@fhanik - I'm not too sure why the build is failing, I haven't touched CAS code... thanks for looking into it for me.
